sql/core/src/main/scala/org/apache/spark/sql/execution/HiveResult.scala

```diff
@@ -37,30 +37,45 @@ object HiveResult {
    * Returns the result as a hive compatible sequence of strings. This is used in tests and
    * `SparkSQLDriver` for CLI applications.
    */
-  def hiveResultString(executedPlan: SparkPlan): Seq[String] = executedPlan match {
-    case ExecutedCommandExec(_: DescribeCommandBase) =>
-      formatDescribeTableOutput(executedPlan.executeCollectPublic())
-    case _: DescribeTableExec =>
-      formatDescribeTableOutput(executedPlan.executeCollectPublic())
-    // SHOW TABLES in Hive only output table names while our v1 command outputs
-    // database, table name, isTemp.
-    case command @ ExecutedCommandExec(s: ShowTablesCommand) if !s.isExtended =>
-      command.executeCollect().map(_.getString(1))
-    // SHOW TABLES in Hive only output table names while our v2 command outputs
-    // namespace and table name.
-    case command : ShowTablesExec =>
-      command.executeCollect().map(_.getString(1))
-    // SHOW VIEWS in Hive only outputs view names while our v1 command outputs
-    // namespace, viewName, and isTemporary.
-    case command @ ExecutedCommandExec(_: ShowViewsCommand) =>
-      command.executeCollect().map(_.getString(1))
-    case other =>
-      val result: Seq[Seq[Any]] = other.executeCollectPublic().map(_.toSeq).toSeq
-      // We need the types so we can output struct field names
-      val types = executedPlan.output.map(_.dataType)
-      // Reformat to match hive tab delimited output.
-      result.map(_.zip(types).map(e => toHiveString(e)))
```

Comment:
@MaxGekk, thinking about what I just wrote to @cloud-fan, would the toHiveString here already handle conversion using the correct session timezone, not the JVM timezone?
Or is there some other case that doesn't work? E.g. about hybrid vs. proleptic calendar?

If that is the case, then we should also set the DATETIME_JAVA8API_ENABLED in the withLocalProperties around Thriftserver JDBC/ODBC operations, to make it work correctly also there.

Reply:
The problem is in `dateFormatter`. Currently, its legacy formatter which is used for `java.sql.Date` doesn't respect to the Spark session time zone and depends on JVM time zone. It works fine for Java 8 `LocalDate` and respect the session time zone.

I tried to fix the issue in https://github.com/apache/spark/pull/28709 but the fix brings more troubles than fixes.

Reply:
I do believe the proper solution is to switch to Java 8 Api.

Reply:
Do date formatting depend on timezone for output formatting?
I thought that timezone is only needed for date parsing, for special cases such as 'now' or 'today' or 'yesterday'?
Or is it the hybrid/proleptic calendar output formatting depend on the timezone?

Reply:
@juliuszsompolski See the comment https://github.com/apache/spark/pull/28687/files#diff-303a0a0e4383242d22307edcdd82e1f1R75-R82 in the PR https://github.com/apache/spark/pull/28687

Reply:
I have tried to make the legacy date formatter independent from the JVM time zone in the PR https://github.com/apache/spark/pull/28709 but it brings some issues too: https://github.com/apache/spark/pull/28709#issuecomment-638123922 . Need to modify more legacy functions to make them independent from the JVM time zone.

Reply:
Following that example:
```
$ export TZ="Europe/Moscow"
$ ./bin/spark-sql -S
spark-sql> set spark.sql.session.timeZone=America/Los_Angeles;
spark.sql.session.timeZone	America/Los_Angeles
spark-sql> select date '2020-06-03';
2020-06-02
spark-sql> select make_date(2020, 6, 3);
2020-06-02
```
Could you explain why does the make_date(2020, 6, 3) -> 2020-06-02 happens?
Does make_date create a date of midnight 2020-6-3 in Moscow TZ, and it gets returned in America/Los_Angeles, where it is still 2020-6-2?
Could you explain step by step with examples what type and what timezone is used during parsing, during collecting, and for the string display before and after the changes?

Reply:
- The date literal '2020-06-03' (and make_date(2020, 6, 3)) is converted to the number of days since the epoch '1970-01-01'. The result is 18416, and it doesn't depend on time zone. You get the same via Java 8 API:
```scala
scala> println(LocalDate.of(2020, 6, 3).toEpochDay)
18416
```
The number is stored as date value internally in Spark.

- To print it out, we should collect it and convert to string. The following steps are for Java 8 OFF:
1. The days are converted to `java.sql.Date` by `toJavaDate()` which is called from https://github.com/apache/spark/blob/b917a6593dc969b9b766259eb8cbbd6e90e0dc53/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystTypeConverters.scala#L306-L309 .
  1. `toJavaDate()` has to create an instance of java.sql.Date from milliseconds since the epoch 1970-01-01 00:00:00Z in UTC time zone. It converts the days 18416 to milliseconds via 18416 * 86400000 and gets 1591142400000.
  2. 1591142400000 is interpreted as local milliseconds in the JVM time zone `Europe/Moscow` which has wall clock offset of 10800000 millis or 3 hours. So, 1591142400000 is shifted by 10800000 to get "UTC timestamp". The result is 1591131600000 which is:
      - `2020-06-02T21:00:00` in UTC
      - `2020-06-03T00:00:00` in Europe/Moscow
      - `2020-06-02T14:00:00` in America/Los_Angeles

2. new Date(1591131600000) is collected and formatted in `toHiveString` by using the legacy date formatter. Currently, the legacy date formatter ignores Spark session time zone `America/Los_Angeles` and uses JVM time zone `Europe/Moscow`. In this way, it converts `new Date(1591131600000)` = `2020-06-03T00:00:00` in Europe/Moscow to `2020-06-03`. Looks fine but after this PR https://github.com/apache/spark/pull/28709, it takes `America/Los_Angeles` and performs the conversion `2020-06-02T14:00:00 America/Los_Angeles` to `2020-06-02`

So, the problem is in `toJavaDate()` which still uses the default JVM time zone.

Reply:
And one more nuance, the legacy type `java.sql.Date` is not local date as Java 8 type `java.time.LocalDate` is. It is actually a timestamp in UTC linked to the JVM time zone. Using it as a local date is not good idea at all but this is Spark's legacy code.

Reply:
So everything is OK now? Enabling jave 8 time API is only for better performance and support negative year?

BTW I doubt if we can support negative year in thriftserver. Even if the server-side can generate the datetime string correctly. The client-side parse the string using `Timestamp.of` which doesn't support negative year.

Reply:
Everything is ok when JVM time zone on executors (where `toJavaDate` is called) is equal to JVM time zone on the driver (where `HiveResult` is initialized). And both JVM time zones are equal to Spark's session time zone.